### PR TITLE
Feat: server boot device selection

### PR DESCRIPTION
**Pull request by cbeti-ionos**
## Checklist

- [x] PR name added as appropriate (e.g. `feat:`/`fix:`/`doc:`/`test:`/`refactor:`)
- [x] Tests added or updated
- [x] Documentation updated
- [x] Changelog updated and version incremented (label: upcoming release)
- [ ] Github Issue linked if any
- [ ] Jira task updated

### CODE REVIEW

File: ionoscloud/resource_server_boot_device_selection.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+				Description: "ID of the first attached volume of the Server, which will be the default boot volume unless another is explicitly specified.",
+				Computed:    true,
+			},
+		},
+
+		Timeouts: &resourceDefaultTimeouts,
+	}
+}
+
+func checkBootDeviceSelectorImmutableFields(_ context.Context, diff *schema.ResourceDiff, _ interface{}) error {
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_server_boot_device_selection.go:56]: Would probably be better off if we set forceNew on the fields that are immutable, like datacenter_id and server_id

File: services/cloudapi/cloudapiserver/server.go [hunk 2]
```diff
@@ -121,6 +179,80 @@ func (ss *Service) UpdateVmState(ctx context.Context, datacenterID, serverID, ne
...
+		return err
+	}
+	if strings.EqualFold(oldBootDeviceID, newBootDeviceID) {
+		return nil
+	}
+
+	newBdType := BootDeviceTypeCDROM
+	_, apiResponse, err := ss.Client.ImagesApi.ImagesFindById(ctx, newBootDeviceID).Execute()
+	if err != nil {
+		if !(apiResponse != nil && apiResponse.StatusCode == 404) {
```
Comment (cristiGuranIonos) [inline services/cloudapi/cloudapiserver/server.go:194]:
```suggestion
		if !(apiResponse != nil && apiResponse.StatusCode == 404) {
```
		if !apiResponse.HttpNotFound() {

File: services/cloudapi/cloudapiserver/server.go [hunk 3]
```diff
@@ -121,6 +179,80 @@ func (ss *Service) UpdateVmState(ctx context.Context, datacenterID, serverID, ne
 	return nil
 }
 
+func (ss *Service) UpdateBootDevice(ctx context.Context, datacenterID, serverID, newBootDeviceID string) error {
```
Comment (cristiGuranIonos) [inline services/cloudapi/cloudapiserver/server.go:182]: add a comment here to explain what happens if a cdrom is previously attached and is moved to a volume

File: services/cloudapi/cloudapiserver/server.go [hunk 4]
```diff
@@ -121,6 +179,80 @@ func (ss *Service) UpdateVmState(ctx context.Context, datacenterID, serverID, ne
...
+			}
+			// detach the cdrom
+			apiResponse, err = ss.Client.ServersApi.DatacentersServersCdromsDelete(ctx, datacenterID, serverID, oldBootDeviceID).Execute()
+			if err != nil {
+				return err
+			}
+			if errState := cloudapi.WaitForStateChange(ctx, ss.Meta, ss.D, apiResponse, schema.TimeoutUpdate); errState != nil {
+				return errState
+			}
+			return nil
```
Comment (cristiGuranIonos) [inline services/cloudapi/cloudapiserver/server.go:216]: add a log here where we mention what/who was detached

File: ionoscloud/resource_server_test.go [hunk 5]
```diff
@@ -805,6 +805,49 @@ func TestAccServerWithLabels(t *testing.T) {
...
+		ProviderFactories: testAccProviderFactories,
+		CheckDestroy:      testAccCheckServerDestroyCheck,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccMultipleBootDevices,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttrPair(constant.ServerResource+"."+constant.ServerTestResource, "boot_volume", constant.ServerResource+"."+constant.ServerTestResource, "inline_volume_ids.0"),
+					resource.TestCheckResourceAttr(constant.ServerResource+"."+constant.ServerTestResource, "boot_cdrom", ""),
+				)},
+			// The server object is updated 'outside' of the resource, so the state of the server resource won't be won't be refreshed in the same step
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_server_test.go:823]:
```suggestion
			// The server object is updated 'outside' of the resource, so the state of the server resource won't be  refreshed in the same step
```

File: services/cloudapi/cloudapiserver/server.go (hunk 4 above)
Comment (cbeti-ionos) [reply services/cloudapi/cloudapiserver/server.go]: refactored the logic a bit and added some extra logging - [90b61bb](https://github.com/ionos-cloud/terraform-provider-ionoscloud/pull/507/commits/90b61bb22669e6c0866d253d4535805c673b7c73)

File: services/cloudapi/cloudapiserver/server.go (hunk 3 above)
Comment (cbeti-ionos) [reply services/cloudapi/cloudapiserver/server.go:186]: solved in - [90b61bb](https://github.com/ionos-cloud/terraform-provider-ionoscloud/pull/507/commits/90b61bb22669e6c0866d253d4535805c673b7c73)

File: services/cloudapi/cloudapiserver/server.go (hunk 2 above)
Comment (cbeti-ionos) [reply services/cloudapi/cloudapiserver/server.go]: solved in - [90b61bb](https://github.com/ionos-cloud/terraform-provider-ionoscloud/pull/507/commits/90b61bb22669e6c0866d253d4535805c673b7c73)

File: ionoscloud/resource_server_test.go (hunk 5 above)
Comment (cbeti-ionos) [reply ionoscloud/resource_server_test.go]: typo fixed in - [6dd8e09](https://github.com/ionos-cloud/terraform-provider-ionoscloud/pull/507/commits/6dd8e09c4de1c678f0c2f24d6ee64e48aa1b5f4a)

**Review by cristiGuranIonos — DISMISSED** [review]

**Review by cristiGuranIonos — APPROVED** [review]

**Pull request merged**